### ISSUE TITLE
What is the difference between DenseVariational and DenseReparameterization?

### ISSUE THREAD

**Issue by nbro** [tensorflow/probability#727]
The (global) re-parametrization trick is described in the paper [Auto-Encoding Variational Bayes](https://arxiv.org/abs/1312.6114). The local re-parametrization trick is described in [Variational Dropout and the Local Reparameterization Trick](https://arxiv.org/abs/1506.02557). The flipout estimator is described in [Flipout: Efficient Pseudo-Independent Weight Perturbations on Mini-Batches](https://arxiv.org/abs/1803.04386). 

The documentation of the [`tfp.layers.DenseVariational`](https://www.tensorflow.org/probability/api_docs/python/tfp/layers/DenseVariational) says

> This layer uses variational inference to fit a "surrogate" posterior to the distribution over both the kernel matrix and the bias terms

> This layer fits the "weights posterior" according to the following generative process:

```
[K, b] ~ Prior()
M = matmul(X, K) + b
Y ~ Likelihood(M)
```

How is this different than variational inference with the re-parametrization trick?  Does mean that, in the case of `tfp.layers.DenseVariational`, you directly sample from the prior? Shouldn't this create high-variance during training?  Moreover, shouldn't you actually sample from the posterior? Finally, what does "likelihood" mean in this case?

**Comment by srvasude** [comment]
@jvdillon who might be able to provide more context.

I don't think these should be different. I think the way this arose is that `DenseVariational` is a more newly written layer that conforms with some of the other Distribution Layers, while the `DenseReparameterization` and `DenseLocalReparameterization` and friends were written further back so may not conform in the same way. The layer itself should use the reparameterization trick (as we do automatically in most TFP code).

The action items here would be to consolidate these list of layers, as well as maybe update some of the older layers (like `DenseLocalReparameterization`).

**Comment by nbro** [comment]
@srvasude If you look at the [source code of these classes](https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/layers/dense_variational.py), you will see that both `DenseReparameterization` and `DenseLocalReparameterization` derive from a basic class called `_DenseVariational`. They only override the method `_apply_variational_kernel`.

I haven't yet looked at the details of `_DenseVariational` and [`DenseVariational`](https://github.com/tensorflow/probability/blob/v0.9.0/tensorflow_probability/python/layers/dense_variational_v2.py), though.

**Comment by hartikainen** [comment]
@nbro did you ever figure out the difference between these two? I'm a little confused about them too.

**Comment by nbro** [comment]
@hartikainen `DenseReparameterization` implements the forward pass by simply sampling from the posterior, without any reparametrization, i.e. it directly samples from the posterior distribution, and it uses this sample to compute the output of layer. I don't know why it's called `DenseReparameterization` because it's not really doing any re-parametrization, as far as I understand. `DenseLocalReparameterization` performs a slightly different sampling operation. You can find the details [here](https://github.com/tensorflow/probability/blob/v0.9.0/tensorflow_probability/python/layers/dense_variational.py#L541). In any case, this is the only difference between the two, i.e. how they sample the kernel during the forward pass (i.e. when `call` is called).

**Comment by hartikainen** [comment]
Thanks @nbro, I think that makes sense! Does the `DenseVariational` (not `_DenseVariational`) work the same way as `DenseReparameterization`, as hinted by @srvasude above?

**Comment by nbro** [comment]
@hartikainen Ops, I had forgotten about [`DenseVariational`](https://github.com/tensorflow/probability/blob/v0.9.0/tensorflow_probability/python/layers/dense_variational_v2.py).

`DenseVariational` is slightly different from the others, in terms of API and functionality. Here are the differences

- `DenseVariational` allows you to weight the KL divergence by specifying this weight as a parameter in the constructor of the class (i.e. `__init__`), while all other Bayesian layers do not allow you to do this, unless you override the function that computes the KL divergence
    - This seems to be the only "advantage"
- `DenseVariational` does not allow you to treat the bias in the same way as the kernel, which means that you can't really compute KL divergence between the prior and posterior biases. By default, `DenseReparameterization` doesn't do this anyway, but you can do that.
- In `DenseVariational`, the function that builds the posterior and priors are not provided by default. This is really weird

**Comment by hartikainen** [comment]
Awesome, now it makes a lot more sense. Feels like these classes should be documented better but for now, hopefully others being confused can just find their way to your answers. Thanks again @nbro!

**Comment by huckiyang** [comment]
Just an add-on. I think the `tfp.layers.DenseVariational` was trying to assign uncertain weights like the [Weight Uncertainty in Neural Networks](https://arxiv.org/pdf/1505.05424.pdf). To write a pure Bayesian NN, we may refer to this [tutorial](https://keras.io/examples/keras_recipes/bayesian_neural_networks/#experiment-2-bayesian-neural-network-bnn).

Just found out both `DenseFlipout` and `DenseReparameterization` are the older layers API as the discussion in https://github.com/tensorflow/probability/issues/359
